dsc/workflows/base/base_workflow.py

```diff
@@ -0,0 +1,199 @@
+from __future__ import annotations
+
+import logging
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Any, final
+
+from dsc.exceptions import InvalidDSpaceMetadataError
+from dsc.item_submission import ItemSubmission
+
+if TYPE_CHECKING:
+    from collections.abc import Iterator
+
+logger = logging.getLogger(__name__)
+
+
+class BaseWorkflow(ABC):
+    """A base workflow class from which other workflow classes are derived."""
+
+    def __init__(
+        self,
+        workflow_name: str,
+        submission_system: str,
+        email_recipients: list[str],
+        metadata_mapping: dict,
+        s3_bucket: str,
+        s3_prefix: str | None,
+        collection_handle: str,
+        output_queue: str,
+    ) -> None:
+        """Initialize base instance.
+
+        Args:
+            workflow_name: The name of the workflow.
+            submission_system: The system to which item submissions will be sent (e.g.
+            DSpace@MIT)
+            email_recipients: The email addresses to notify after runs of the workflow.
+            metadata_mapping: A mapping file for generating DSpace metadata from the
+            workflow's source metadata.
+            s3_bucket: The S3 bucket containing bitstream and metadata files for the
+            workflow.
+            s3_prefix: The S3 prefix used for objects in this workflow. This prefix does
+            NOT include the bucket name.
+            collection_handle: The handle of the DSpace collection to which submissions
+            will be uploaded
+            output_queue: The SQS output queue used for retrieving result messages from
+            the workflow's submissions.
+        """
+        self.workflow_name: str = workflow_name
+        self.submission_system: str = submission_system
+        self.email_recipients: list[str] = email_recipients
+        self.metadata_mapping: dict = metadata_mapping
+        self.s3_bucket: str = s3_bucket
+        self.s3_prefix: str | None = s3_prefix
+        self.collection_handle: str = collection_handle
+        self.output_queue: str = output_queue
+
+    @final  # noqa: B027
+    def run(self) -> None:
+        """Run workflow to submit items to  the DSpace Submission Service.
+
+        PLANNED CODE:
+
+        sqs_client = SQSClient()
+        for item_submission in self.item_submissions_iter():
+            item_submission.upload_dspace_metadata(
+                self.s3_bucket, item_submission.metadata_keyname
+            )
+            sqs_client.send_submission_message(
+                item_submission.item_identifier,
+                self.workflow_name,
+                self.output_queue,
+                self.submission_system,
+                self.collection_handle,
+                item_submission.metadata_uri,
+                item_submission.bitstream_uris,
+            )
```

Comment:
I'm happy to wait on this discussion if it would be helpful to tease out when this is actually built out.

Given this proposal, it feels like it could be a little difficult to test with this approach, where both `upload_space_metadata()` and `send_submission_message()` appear to be crafting objects _before_ they send them off into the ether.

What about a more verbose, but potentially easier to test flow:

```python
for item_submission in self.item_submissions_iter():

    # prepare dspace metadata and upload to S3
    dspace_metadata = item_submission.prepare_dspace_metadata()
    self.upload_dspace_metadata(metadata_s3_key, dspace_metadata)
    
    # send SQS message to DSS input queue
    submission_sqs_message = item_submission.prepare_sqs_input_message()
    sqs_client.publish_submission_message(submission_sqs_message)
```

This makes some assumptions that you may or may not agree with:
- the `BaseWorkflow` class would have the functionality to perform uploads to S3
- the `ItemSubmission` class could only prepare _payloads_ that will get sent
  - this removes the need for `ItemSubmission` to know anything about where they will go (S3 buckets, queues, 
etc.; all of which are sort of set on the workflow instance itself)
- the `ItemSubmission` class is what prepares SQS messages, but the `SQSClient` actually sends it
  - or, perhaps the `BaseWorkflow` could have a method like `submit_to_dss_input_queue()` which itself would utilize the `SQSClient` for actually publishing the message, but it would still be the _workflow_ that kicked that off and was ultimately responsible for it

Just food for thought.  But I realize this just a stub, and will get more attention when that part is written.

Reply:
This is very helpful, things are already shifting from this proposed code and testing was a major consideration. This will be ready for full review in the next PR after I refactor the SQS functionality!